### ISSUE TITLE
Do not block while lazily loading networks, building distance tables, and linking

### ISSUE THREAD

**Issue by abyrd** [conveyal/r5#381]
Right now some very slow operations are done lazily as needed. The best example is loading a graph. When the worker receives a task on a graph it does not have loaded, it accepts the task, then loads or even builds the entire graph as part of handling the request. Even if loading graphs can be made very fast, at least building graphs should be a separate task, not something that may or may not happen as part of each request.

A huge problem with this is that HTTP connections from the UI are held open for very long periods of time with no feedback as to what's happening from the backend or worker. See #447.

Like loading and building networks, building distance tables and linking grids to the street network is done lazily as needed. Workers receive requests which block for a long time waiting for these actions to finish. If the actions take a very long time the front end effectively locks up and the request eventually times out. This is bad user experience.

Although this is harmless for batch jobs which happen asynchronously, for single point jobs this should be handled the same way as cluster startup: we should return a "service temporarily unavailable" code and the client should retry later.

This implies that single point work would no longer use the current system where the worker pulls the requests, and instead the workers would handle single point work by being contacted by the broker. The broker/backend would then just be proxying requests to a particular worker, and it's easy to find an appropriate worker using the worker catalog.

**Comment by abyrd** [comment]
The last paragraph above has now been implemented - the broker proxies requests to a worker chosen from the catalogue. This enables immediate responses to all HTTP requests stating whether the worker is ready, and asynchronous linking and distance table construction. Note that the distance table construction is actually the slowest operation and responsible for serious lockups.

**Comment by abyrd** [comment]
I've done some work on this, creating an asynchronous loading cache to replace the Guava loadingCache and ConcurrentHashMap.putIfAbsent. I checked into library options first -  Guava does not cleanly support such cases: https://github.com/google/guava/issues/1350

In its current form I've got more of an async lazy-loading concurrent map than a cache because it doesn't have eviction (yet). Eviction caused us all kinds of problems before and anyway once a reference to the cached value escapes to the caller, it's really hard/impossible in Java to ensure that it is out of memory before another one takes its place. So for now we'll just be letting things accumulate in memory and try to make sure the workers only work on a small enough domain that they don't get full.

Observations once I got to implementing a linkage cache: We need to be able to select blocking / async, or have two different cache implementations, one blocking, one async. We don't actually want most of these lazy cache fetches to be asynchronous, unless we've got some kind of sophisticated event driven task dependency tree thing going on that allows tasks to suspend and resume. The reason is that we fetch from lazy caches recursively - the lazy network cache then hits a lazy distance table cache, and linkage caches hit other linkage caches for their parent street layers. When the method for building/loading a value into a cache completes, that usually means the value is ready to put into the cache. But since those methods need to recursively hit caches, in a completely async style they would exit early. It is also noteworthy that things need to happen in a particular order: distance tables and linkages are dependent on already having the network in hand with scenario applied.

An approach that makes more sense is to make all the cache value loading synchronous/single threaded, but wrap the whole thing in an asynchronous "cache primer" / pre-loader that checks all the preconditions right up front, before we begin handling an incoming task, and enqueues one big synchronous task to pull all the right data into memory if everything is not already there. Concentrating this in one place also makes it easy to switch between blocking and nonblocking approaches (for regional versus single-point tasks) by either running the cache-priming method directly in the calling thread, or by skipping it entirely and letting the task handlers block while they are working on the task (as we currently do in regional analyses, with no real ill effects).

The keys for this pre-loader should be: (scenarioId, grid, modes) and it would fall through to normal task handling only when all needed data were already available.

Places to look:
com.conveyal.r5.analyst.TravelTimeComputer#computeTravelTimes() // Linkages are loaded here for access, egress, and direct modes
com.conveyal.r5.analyst.cluster.AnalystWorker#handleOneRequest() // Network is loaded here
Distance tables are built when applying scenarios, in the TransportNetworkCache, which is called in handleOneRequest.
